ATTESTATION_VERIFICATION_PROPOSAL.md

```diff
@@ -0,0 +1,253 @@
+      }
+    }
+  }
+}
+```
+We are passing the `images` variable which has `containers`, `initContainers` and `ephimeralContainers` in it, containing information of all the images used.
+
+Current Response Format:
+```json
+{
+  "verified": true,
+  "message": "...",
+  "results": [
+      {
+         "name": "container1",
+         "path":  "/spec/containers/0",
+         "image":  "ghcr.io/kyverno/test-verify-image@sha256:b31bfb4d0213f254d361e0079deaaebefa4f82ba7aa76ef82e90b4935ad5b105"
+      } 
+  ]
+}
+```
+It returns a verified `boolean` that indicated whether verification was successful or not, `message` which is an optional field, and `results` which is a list of JSONPatch compatible objects containing entries for every image.
+
+We should add an array to the `requestData` called `attestationType` as follows
+```json
+{
+  "images": {
+    "containers": {
+      "tomcat": {
+        "registry": "https://ghcr.io",
+        "path": "tomcat",
+        "name": "tomcat",
+        "tag": "9",
+        "jsonPointer": "spec/container/0/image"
+      }
+    },
+    "initContainers": {
+      "vault": {
+        "registry": "https://ghcr.io",
+        "path": "vault",
+        "name": "vault",
+        "tag": "v3",
+        "jsonPointer": "spec/initContainer/0/image"
+      }
+    }
+  },
+  "attestationType": [
+    "application/vnd.cyclonedx",
+    "sbom/cyclone-dx",
+    "application/sarif+json"
+  ]
+}
+```
+Here is how it will look in the policy
+```yaml
+context:
+- name: response
+  apiCall:
+    method: POST
+    data:
+    - key: images
+      value: "{{ images }}"
+    - key: attestationType
+      value: |-
+        - application/vnd.cyclonedx
+        - sbom/cyclone-dx
+        - application/sarif+json
+    service:
+      url: https://svc.kyverno-notation-aws/checkimages
+      caBundle: |-
+...
+```
+Here is a gist showing the entire policy: https://gist.github.com/Vishal-Chdhry/ec90442cf892c4d7db169ff45918615d
+
+For every image we will check whether it has any or all of the attestations attached to it. When the attestations is found, we will return the following:
+```json
+{
+  "verified": true,
+  "message": "...",
+  "results": [
+      {
+         "name": "container1",
+         "path":  "/spec/containers/0",
+         "image":  "ghcr.io/kyverno/test-verify-image@sha256:b31bfb4d0213f254d361e0079deaaebefa4f82ba7aa76ef82e90b4935ad5b105",
+         "attestations": [
+          {
+            "type": "sbom/cyclone-dx",
+            "digest": "sha256:b31bfb4d0213f254d361e0079deaaebefa4f82ba7aa76ef82e90b4935ad5b105",
+            "payload": "the entire SBOM"
+          }
+         ]
+      } 
+  ]
+}
+``` 
+
+For validation using this result, the policy should look like this
+```yaml
+validate:
+  message: "not allowed"
+  foreach:
+    - list: response.results
+      deny:
+        conditions:
+          any:
+          - key: application/vnd.cyclonedx # To shortcircuit the check if the type is not present in the image
```

Comment:
Not sure I follow this....will the check succeed or fail if the attestation is not present. It should fail right? The comment makes it seem like the check will be skipped.

Reply:
I am thinking about changing it to 
```yaml
validate:
  message: "not allowed"
  foreach:
    - list: response.attestations[?type= 'sbom/cyclone-dx'].payload
      deny:
        conditions:
          - all:
            - key: "{{ element.components[].licenses[].expression }}"
              operator: AllNotIn
              value: ["GPL-2.0", "GPL-3.0"]
```
Result form the api should be of format:
```json
{
  "verified": true,
  "message": "...",
  "results": [
      {
         "name": "container1",
         "path":  "/spec/containers/0",
         "image":  "ghcr.io/kyverno/test-verify-image@sha256:b31bfb4d0213f254d361e0079deaaebefa4f82ba7aa76ef82e90b4935ad5b105"
      } 
  ],
  "attestations": [
      {
        "type": "sbom/cyclone-dx",
        "digest": "sha256:b31bfb4d0213f254d361e0079deaaebefa4f82ba7aa76ef82e90b4935ad5b105",
        "payload": "the entire SBOM",
        "path":  "/spec/containers/0",
        "image":  "ghcr.io/kyverno/test-verify-image@sha256:b31bfb4d0213f254d361e0079deaaebefa4f82ba7aa76ef82e90b4935ad5b105"
      }
    ]
}
```